workflow/controller/operator.go

```diff
@@ -2630,6 +2650,38 @@ func (woc *wfOperationCtx) executeResource(ctx context.Context, nodeName string,
 	return node, err
 }
 
+func (woc *wfOperationCtx) executeData(ctx context.Context, nodeName string, templateScope string, tmpl *wfv1.Template, orgTmpl wfv1.TemplateReferenceHolder, opts *executeTemplateOpts) (*wfv1.NodeStatus, error) {
+	node := woc.wf.GetNodeByName(nodeName)
+	if node == nil {
+		node = woc.initializeExecutableNode(nodeName, wfv1.NodeTypePod, templateScope, tmpl, orgTmpl, opts.boundaryID, wfv1.NodePending)
+	} else if !node.Pending() {
+		return node, nil
+	}
+
+	if tmpl.Data.UsePod() || woc.controller.Config.WorkflowRestrictions.MustUsePodWithDataTemplate() {
+		dataTemplate, err := json.Marshal(tmpl.Data)
+		if err != nil {
+			return node, fmt.Errorf("could not marhsal data in transformation: %w", err)
+		}
+
+		mainCtr := woc.newExecContainer(common.MainContainerName, tmpl)
+		mainCtr.Command = []string{"argoexec", "data", string(dataTemplate)}
+		_, err = woc.createWorkflowPod(ctx, nodeName, *mainCtr, tmpl, &createWorkflowPodOpts{onExitPod: opts.onExitTemplate, executionDeadline: opts.executionDeadline, includeScriptOutput: true})
+		if err != nil {
+			return woc.requeueIfTransientErr(err, node.Name)
+		}
+	} else {
+		transformedData, err := data.ProcessData(tmpl.Data, newOperatorDataSourceProcessor())
+		if err != nil {
+			errorMessage := fmt.Sprintf("could not process data: %s", err)
+			return woc.markNodePhase(node.Name, wfv1.NodeFailed, errorMessage), fmt.Errorf(errorMessage)
+		}
+		node = woc.finalizeTransformNode(nodeName, transformedData)
```

Comment:
inline this func?

Reply:
I like it on its own function, helps keep this already bloated function more simple